### PR TITLE
[MLIR][Torch] Add TorchToLinalg lowering for AtenAvgPool3dOp

### PR DESCRIPTION
**Pull request by vivekkhandelwal1**
This commit also fixes the average pool op' test failing for OnnxToLinalg lowering.

Signed-Off By: Vivek Khandelwal <vivekkhandelwal1424@gmail.com>

### CODE REVIEW

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 1]
```diff
@@ -527,6 +541,95 @@ template <typename OpTy, typename PoolingOpTy, int Dim>
...
+        {rewriter.getAffineDimExpr(0), rewriter.getAffineDimExpr(1),
+         rewriter.getAffineDimExpr(2), rewriter.getAffineDimExpr(3),
+         rewriter.getAffineDimExpr(4)},
+        context);
+    auto iteratorTypes =
+        SmallVector<utils::IteratorType>(5, utils::IteratorType::parallel);
+    iteratorTypes.append(3, utils::IteratorType::reduction);
+    SmallVector<AffineMap> indexingMaps = {mapInput, mapKernel, mapOutput};
+
+    result = rewriter
```
Comment (rsuderman) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:615]: Is there a reason you can't use the linalg pooling operation:
https://github.com/shark-infra/llvm-project/blob/3d50e75db58dfd2ef6c35a588ed4e5b03ce53efa/mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml#L5563

**Review by rsuderman — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToLinalg/Pooling.cpp (hunk 1 above)
Comment (vivekkhandelwal1) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:615]: In order to use that we have to change the ordering of input tensor values as well as the result tensor value. 

Also, I have followed what has been done before for the MaxPool3D op.

Comment (rsuderman) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:615]: It is still preferable to use the linalg specializations. There are a lot of optimizations for these kinds of operations that we won't get for pooling.

**Comment by AmosLewis** [comment]
You will need this commit to
[ONNX] Fix padding attributes for onnx.AveragePool https://github.com/llvm/torch-mlir/commit/eb3b62f966f69df5e434b5c4cb2663303c1d68bb

**Comment by AmosLewis** [comment]
Could you also test with model https://github.com/nod-ai/SHARK-TestSuite/issues/185 in shark-testsuites with `--torchtolinag` flag before iree bump:
`python ./run.py --torchmlirbuild ../../torch-mlir/build --tolerance 0.001 0.001 --cachedir ./huggingface_cache --ireebuild ../../iree-build  -f onnx -g models --mode onnx --report --tests onnx/models/VideoResNet_vaiq_int8 --torchtolinalg`

**Comment by vivekkhandelwal1** [comment]
> Could you also test with model [nod-ai/SHARK-TestSuite#185](https://github.com/nod-ai/SHARK-TestSuite/issues/185) in shark-testsuites with `--torchtolinag` flag before iree bump: `python ./run.py --torchmlirbuild ../../torch-mlir/build --tolerance 0.001 0.001 --cachedir ./huggingface_cache --ireebuild ../../iree-build -f onnx -g models --mode onnx --report --tests onnx/models/VideoResNet_vaiq_int8 --torchtolinalg`

It fails during iree-compile with failure for 3d conv but passed for torch-mlir.

**Comment by vivekkhandelwal1** [comment]
> You will need this commit to [ONNX] Fix padding attributes for onnx.AveragePool [eb3b62f](https://github.com/llvm/torch-mlir/commit/eb3b62f966f69df5e434b5c4cb2663303c1d68bb)

We can do this in a seperate PR, since the current PR doesn't make any change in the Onnx implementation. Or, you can create a PR with your commit for the same.

**Comment by vivekkhandelwal1** [comment]
> Could you also test with model [nod-ai/SHARK-TestSuite#185](https://github.com/nod-ai/SHARK-TestSuite/issues/185) in shark-testsuites with `--torchtolinag` flag before iree bump: `python ./run.py --torchmlirbuild ../../torch-mlir/build --tolerance 0.001 0.001 --cachedir ./huggingface_cache --ireebuild ../../iree-build -f onnx -g models --mode onnx --report --tests onnx/models/VideoResNet_vaiq_int8 --torchtolinalg`

I'm getting this:
```
Status report for run: test-onnx using mode:onnx todtype:default backend:llvm-cpu

| tests                             | model-run   | onnx-import   | torch-mlir   | iree-compile   | inference   |
|:----------------------------------|:------------|:--------------|:-------------|:---------------|:------------|
| onnx/models/VideoResNet_vaiq_int8 | passed      | failed        | notrun       | notrun         | notrun      |

```

**Review by rsuderman — CHANGES_REQUESTED** [review]

**Comment by vivekkhandelwal1** [comment]
Hi @rsuderman, can you please re-review this PR?

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 2]
```diff
@@ -173,11 +173,44 @@ static LogicalResult createPoolingOp(
...
-               .create<OpTy>(loc, outTensorInitialized.getType(),
-                             ValueRange{paddedInput, windowTensor},
-                             outTensorInitialized, stridesAttr, dilationAttr)
-               .getResult(0);
+  SmallVector<int64_t> dimensions;
+  Value permutedInput = paddedInput, permutedOutput = outTensorInitialized;
+  if (dimensionality == 3) {
+    // Permute input and output tensor as follows:
+    // (n,c,d,h,w) -> (n,d,h,w,c)
+    dimensions = {0, 2, 3, 4, 1};
```
Comment (rsuderman) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:181]: Declare these in here. The values do not appeared to be used elsewhere so now point to have it declared outside of the if condition.

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 3]
```diff
@@ -173,11 +173,44 @@ static LogicalResult createPoolingOp(
...
+          .create<OpTy>(loc, permutedOutput.getType(),
+                        ValueRange{permutedInput, windowTensor}, permutedOutput,
+                        stridesAttr, dilationAttr)
+          .getResult(0);
+
+  result = poolingResult;
+  if (dimensionality == 3) {
+    // Permute output tensor as follows:
+    // (n,d,h,w,c) -> (n,c,d,h,w)
+    dimensions.clear();
```
Comment (rsuderman) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:206]: See comment about declaration

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 4]
```diff
@@ -611,15 +644,18 @@ class ConvertAtenAvgPoolOp : public OpConversionPattern<OpTy> {
...
-          loc, kernelSizeIntValues[0], kernelSizeIntValues[1]);
-      divisor = isa<Torch::NoneType>(op.getDivisorOverride().getType())
-                    ? kHtimeskW
-                    : adaptor.getDivisorOverride();
-    } else {
-      divisor = kernelSizeIntValues[0];
+    // }
+
+    Value divisor = kernelSizeIntValues[0];
+    for (unsigned i = 1; i < kernelSizeIntValues.size(); i++) {
```
Comment (rsuderman) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:650]: Use bit length specific version. `unsigned` has varying behavior depending on computer.

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 5]
```diff
@@ -611,15 +644,18 @@ class ConvertAtenAvgPoolOp : public OpConversionPattern<OpTy> {
...
+      divisor =
+          rewriter.create<arith::MulIOp>(loc, divisor, kernelSizeIntValues[i]);
+    }
+    if constexpr (!std::is_same<OpTy, AtenAvgPool1dOp>()) {
```
Comment (rsuderman) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:654]: You can just drop the `constexpr` part. The compiler will optionally remove anyway.

**Review by rsuderman — CHANGES_REQUESTED** [review]

**Comment by vivekkhandelwal1** [comment]
Hi @rsuderman, can you please re-review this?

File: lib/Conversion/TorchToLinalg/Pooling.cpp (hunk 5 above)
Comment (vivekkhandelwal1) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:645]: I dropped that but it resulted in an error hence kept it. What should I do?

**Comment by vivekkhandelwal1** [comment]
> Hi @rsuderman, can you please re-review this?

Hi @rsuderman, just a gentle reminder to review this PR.

**Review by rsuderman — APPROVED** [review]

**Pull request merged**